examples/src/keyring/aws_kms/discovery_decrypt.py

```diff
@@ -0,0 +1,78 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""
+When you give the KMS keyring specific key IDs it will use those CMKs and nothing else.
+This is true both on encrypt and on decrypt.
+However, sometimes you need more flexibility on decrypt,
+especially if you might not know beforehand which CMK was used to encrypt a message.
+To address this need, the KMS keyring also supports "discovery" mode.
+In discovery mode, the KMS keyring will do nothing on encrypt
+but will attempt to decrypt *any* data keys that were encrypted under a KMS CMK.
+
+This example shows how to configure and use a KMS keyring in discovery mode.
+
+https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/choose-keyring.html#use-kms-keyring
+
+For an example of how to use the KMS keyring with CMKs in multiple regions,
+see the ``keyring/aws_kms/multiple_regions`` example.
+
+For examples of how to use the KMS keyring with custom client configurations,
+see the ``keyring/aws_kms/custom_client_supplier``
+and ``keyring/aws_kms/custom_kms_client_config`` examples.
+
+For examples of how to use the KMS keyring in discovery mode on decrypt,
+see the ``keyring/aws_kms/discovery_decrypt_in_region_only``
+and ``keyring/aws_kms/discovery_decrypt_with_preferred_region`` examples.
+"""
+import aws_encryption_sdk
+from aws_encryption_sdk.keyrings.aws_kms import KmsKeyring
+
+
+def run(aws_kms_cmk, source_plaintext):
+    # type: (str, bytes) -> None
+    """Demonstrate an encrypt/decrypt cycle using a KMS keyring with a single CMK.
+
+    :param str aws_kms_cmk: The ARN of an AWS KMS CMK that protects data keys
+    :param bytes source_plaintext: Plaintext to encrypt
+    """
+    # Prepare your encryption context.
+    # https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#encryption-context
+    encryption_context = {
+        "encryption": "context",
+        "is not": "secret",
+        "but adds": "useful metadata",
+        "that can help you": "be confident that",
+        "the data you are handling": "is what you think it is",
+    }
+
+    # Create the keyring that determines how your data keys are protected.
+    encrypt_keyring = KmsKeyring(generator_key_id=aws_kms_cmk)
+
+    # Create the KMS discovery keyring that we will use on decrypt.
+    #
+    # Because we do not specify any key IDs, this keyring is created in discovery mode.
+    decrypt_keyring = KmsKeyring()
```

Comment:
I feel like that would be better shown in the client supplier example, demonstrating that if the discovery keyring's client supplier cannot supply a client for a region then the keyring cannot decrypt EDKs from that region.

Reply:
As discussed in https://github.com/awslabs/aws-encryption-sdk-specification/issues/73, the failure mode for any keyring on decrypt is the same. Rather than demonstrate how to handle failure for every example, I think it might be better to have a separate example that only talks about how to correctly handle a failure to decrypt. Thoughts?

Reply:
> I think it might be better to have a separate example that only talks about how to correctly handle a failure to decrypt.

I think this is a super great idea!

Reply:
#229